### ISSUE TITLE
[TASK] Explore a pseudo repartition instead of sort for key grouped operations

### ISSUE THREAD

**Issue by revans2** [NVIDIA/spark-rapids#2863]
**Is your feature request related to a problem? Please describe.**
We have a number of out of core algorithms, aggregations and joins, that we are working on that rely on sort when the data is too large to fit into a single batch.

Sorting data is rather slow. This is especially true when there are more than one key or the key is variable length. It would be nice to explore alternative algorithms that do not involve sorting the data in order to group similar keys into corresponding batches. For aggregations and many join types the data can be hash partitioned before being operated on. Spark does this to distribute the data to multiple tasks to process the data in parallel.  If we run into a situation where the data is too large to fit into memory after this initial partitioning we should look at running similar partition code again over the data to split it up into even smaller groups, which we can then operate on.  If needed this process could be repeated again if the sizes are still too large. The main drawback to this compared to sorting is that we need a way to know when to stop in the face of data skew. 

As such it would be nice to do a prototype of this and see how the performance compares to sorting. If we see a lot of performance potential then we can look at working with cudf to figure out the right way to stop when there is a lot of data skew.

When doing the performance comparison we should also pay attention to aggregations that are sort based in cudf. If the data is already sorted, then cudf will not need to sort it again. In those cases we might lose all of our gains from trying to avoid sorting.
